sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs

```diff
@@ -0,0 +1,103 @@
+            string identityEndpoint = EnvironmentVariables.IdentityEndpoint;
+            string identityHeader = EnvironmentVariables.IdentityHeader;
+            string identityServerThumbprint = EnvironmentVariables.IdentityServerThumbprint;
+
+            if (string.IsNullOrEmpty(identityEndpoint) || string.IsNullOrEmpty(identityHeader) || string.IsNullOrEmpty(identityServerThumbprint))
+            {
+                return default;
+            }
+
+            Uri endpointUri;
+            try
+            {
+                endpointUri = new Uri(identityEndpoint);
+            }
+            catch (FormatException ex)
+            {
+                throw new AuthenticationFailedException(IdentityEndpointInvalidUriError, ex);
+            }
+
+            var pipeline = options.Pipeline;
+
+            if (!options.PreserveTransport)
+            {
+                var customSslHttpPipline = HttpPipelineBuilder.Build(new TokenCredentialOptions { Transport = GetServiceFabricMITransport() });
+
+                pipeline = new CredentialPipeline(pipeline.AuthorityHost, customSslHttpPipline, pipeline.Diagnostics);
+            }
+
+            return new ServiceFabricManagedIdentitySource(pipeline, endpointUri, identityHeader, options.ClientId);
+        }
+
+        internal static HttpClientTransport GetServiceFabricMITransport()
+        {
+            var httpHandler = new HttpClientHandler();
+
+            httpHandler.ServerCertificateCustomValidationCallback = ValidateMsiServerCertificate;
+
+            return new HttpClientTransport(httpHandler);
+        }
+
+        private ServiceFabricManagedIdentitySource(CredentialPipeline pipeline, Uri endpoint, string identityHeaderValue, string clientId) : base(pipeline)
+        {
+            _endpoint = endpoint;
+            _identityHeaderValue = identityHeaderValue;
+            _clientId = clientId;
+        }
+
+        protected override Request CreateRequest(string[] scopes)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = Pipeline.HttpPipeline.CreateRequest();
+
+            request.Method = RequestMethod.Get;
+            request.Headers.Add("secret", _identityHeaderValue);
+            request.Uri.Reset(_endpoint);
+            request.Uri.AppendQuery("api-version", ServiceFabricMsiApiVersion);
+            request.Uri.AppendQuery("resource", resource);
+
+            if (!string.IsNullOrEmpty(_clientId))
+            {
+                request.Uri.AppendQuery("client_id", _clientId);
+            }
+
+            return request;
+        }
+
+        private static bool ValidateMsiServerCertificate(HttpRequestMessage message, X509Certificate2 cert, X509Chain certChain, SslPolicyErrors policyErrors)
+        {
+            // Do any additional validation here
+            if (policyErrors == SslPolicyErrors.None)
+            {
+                return true;
```

Comment:
So if we trust the cert, we don't care whether it's the one we expect?

Reply:
This validation callback came from the sample provided by the service team, https://docs.microsoft.com/en-us/azure/service-fabric/how-to-managed-identity-service-fabric-app-code#acquiring-an-access-token-using-c. I can reach out to them to see if removing this is makes sense.